### PR TITLE
Implemented PMI masking for T5-Pretraining.

### PR DESCRIPTION
**Pull request by OfekGlick**
Changes include adding the PMI vocabulary, n-grams range from size 2-5, Implementing PMI masking in the Data Collator,
Adding relevant flags in the flag file, and unit testing for the PMI masking methods.

### CODE REVIEW

File: corruption_t5.py [hunk 1]
```diff
@@ -39,6 +43,139 @@ def shift_tokens_right(
     return shifted_input_ids
 
 
+def merge_segments(offset, n_grams, covered_indices, whole_words, start_inds, segments_to_merge, ngrams_vocab_set):
```
Comment (zbambergerNLP) [inline corruption_t5.py]: Missing docstring and type annotations.

File: corruption_t5.py [hunk 2]
```diff
@@ -39,6 +43,139 @@ def shift_tokens_right(
...
+    possible_merges = []
+    for i in range(len(start_inds) - n_grams):
+        segment = whole_words[start_inds[i]:start_inds[i + n_grams] - 1]
+        if segment in ngrams_vocab_set:
+            possible_merges.append(list(range(offset + i, offset + i + n_grams)))
+
+    random.shuffle(possible_merges)
```
Comment (zbambergerNLP) [inline corruption_t5.py:73]: nit: Are you sure that given random.setseed() performed earlier , that this call will be deterministic?

File: corruption_t5.py (hunk 1 above)
Comment (zbambergerNLP) [inline corruption_t5.py]: Add a TODO: vectorize this function with pytorch/numpy

File: corruption_t5.py [hunk 3]
```diff
@@ -39,6 +43,139 @@ def shift_tokens_right(
...
+    for seg_inds in possible_merges:
+        if len(set(seg_inds).intersection(covered_indices)) > 0:
+            continue
+        covered_indices.update(seg_inds)
+        segments_to_merge.append(seg_inds)
+
+
+def pmi_word_mask(input_tokens, pmi_vocab, max_predictions=512, mlm_probability=0.5):
```
Comment (zbambergerNLP) [inline corruption_t5.py]: TODO: vectorize this function with pytorch or numpy

File: corruption_t5.py [hunk 4]
```diff
@@ -39,6 +43,139 @@ def shift_tokens_right(
...
+        input_tokens: A tensor of tokens.
+        pmi_vocab: The set of ngrams to use for PMI-based corruption.
+        max_predictions: The maximum number of tokens to mask.
+        mlm_probability: The probability of masking a token.
+    """
+    whole_words_indexes = []
+    whole_words_lists = [[]]
+    whole_words = whole_words_lists[0]
+    for (i, token) in enumerate(input_tokens):
+        if token == '</s>' or token == '<pad>':
```
Comment (zbambergerNLP) [inline corruption_t5.py]: '\<\/s\>' should be a constant and same with '\<pad\>'. Avoid comparing against naked strings.

File: corruption_t5.py [hunk 5]
```diff
@@ -39,6 +43,139 @@ def shift_tokens_right(
...
+    for seg_inds in possible_merges:
+        if len(set(seg_inds).intersection(covered_indices)) > 0:
+            continue
+        covered_indices.update(seg_inds)
+        segments_to_merge.append(seg_inds)
+
+
+def pmi_word_mask(input_tokens, pmi_vocab, max_predictions=512, mlm_probability=0.5):
+    """
+    Create a mask for PMI-based corruption for a sample.
```
Comment (zbambergerNLP) [inline corruption_t5.py:88]: Add another line or two which describe the high-level logic behind creating the masks.

File: corruption_t5.py (hunk 3 above)
Comment (zbambergerNLP) [inline corruption_t5.py]: Missing type annotations

File: corruption_t5.py [hunk 6]
```diff
@@ -39,6 +43,139 @@ def shift_tokens_right(
...
+            whole_words_indexes.append([i])
+            whole_words.append(token.strip("▁").lower())
+    offset = 0
+    covered_indices = set()
+    segments_to_merge = []
+    for whole_words in whole_words_lists:
+        if len(whole_words) == 0:
+            continue
+        added_offset = len(whole_words)
+        whole_words = ' '.join(whole_words)
```
Comment (zbambergerNLP) [inline corruption_t5.py:125]: What about whole words that are separated by linebreaks or other forms of spacing?

File: corruption_t5.py [hunk 7]
```diff
@@ -39,6 +43,139 @@ def shift_tokens_right(
...
+        offset += added_offset
+    segments_to_merge.extend([i] for i in set(range(len(whole_words_indexes))).difference(covered_indices))
+
+    candidates = []
+    for seg_to_merge in segments_to_merge:
+        candidates.append(sum([whole_words_indexes[i] for i in seg_to_merge], []))
+
+    # whole_words_indexes is a list of lists of ints, each list is the
+    # indices part of the whole word to be masked, i.e. the segment to be considered for masking
+    random.shuffle(candidates)
```
Comment (zbambergerNLP) [inline corruption_t5.py:146]: Here too we should be clear that this is can be deterministic (by setting a seed earlier in the code)

File: corruption_t5_test.py
Comment (zbambergerNLP) [inline corruption_t5_test.py:1]: Need new and robust tests for each of the new functions.

File: fine_tune_t5.py [hunk 8]
```diff
@@ -20,6 +20,8 @@
     # Parse flags
     parser = HfArgumentParser((flags.ModelArguments, flags.DataTrainingArguments, flags.TrainingArguments))
     model_args, data_args, training_args = parser.parse_args_into_dataclasses()
+
+    os.environ["WANDB_API_KEY"] = '0bc4005546f2e4d5791ff50a0eeb476ee687d5c6'
```
Comment (zbambergerNLP) [inline fine_tune_t5.py]: Remove from code. This is a public repo.

File: flags.py [hunk 9]
```diff
@@ -102,7 +102,7 @@ class TrainingArguments:
         default=42, metadata={"help": "The seed to use for reproducible training."}
     )
     deepspeed: bool = field(
-        default=True, metadata={"help": "Whether to use deepspeed for training."}
+        default=False, metadata={"help": "Whether to use deepspeed for training."}
```
Comment (zbambergerNLP) [inline flags.py]: Change back to True

File: flags.py [hunk 10]
```diff
@@ -241,11 +245,11 @@ class DataTrainingArguments:
         },
     )
     pre_training_dataset_paths: Optional[str] = field(
-        default="wikipedia,bookcorpus",
+        default="bookcorpus",
```
Comment (zbambergerNLP) [inline flags.py]: Change this back to include wikipedia as well

File: flags.py [hunk 11]
```diff
@@ -210,7 +214,7 @@ class DataTrainingArguments:
         metadata={"help": "The directory where the tokenized datasets will be saved."},
     )
     percent_of_dataset: Optional[int] = field(
-        default=100,
+        default=2,
```
Comment (zbambergerNLP) [inline flags.py]: Change this back to 100 percent.

File: flags.py [hunk 12]
```diff
@@ -241,11 +245,11 @@ class DataTrainingArguments:
         },
     )
     pre_training_dataset_paths: Optional[str] = field(
-        default="wikipedia,bookcorpus",
+        default="bookcorpus",
         metadata={"help": "The name of the dataset to use for pre-training (via the datasets library)."}
     )
     pre_training_dataset_names: Optional[str] = field(
-        default="20220301.en,",  # Only wikipedia has a name, bookcorpus is just a path.
+        default="",  # Only wikipedia has a name, bookcorpus is just a path.
```
Comment (zbambergerNLP) [inline flags.py]: Change back

File: pmi_corruption_test.py
Comment (zbambergerNLP) [inline pmi_corruption_test.py:1]: Merge with existing t5_corruption tests.

File: pmi_corruption_test.py [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+
+
+class PMICorruptionTest(parameterized.TestCase):
+
+    def setUp(self):
+        self.tokenizer = transformers.T5Tokenizer.from_pretrained('google/t5-v1_1-small')
+        random.seed(42)
+
+    @parameterized.named_parameters(
+        *[{
```
Comment (zbambergerNLP) [inline pmi_corruption_test.py]: Do not use this notation, you should have a defined dictionary here with clear testcase names, and potential modifications to parameters that are critical to the function under test.

Comment (zbambergerNLP) [reply pmi_corruption_test.py]: See the examples in the existing test file, corruption_t5_test.py

File: pmi_corruption_test.py [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+            mlm_probability,)
+        self.assertIsNotNone(mask_labels_for_sample)
+
+    @parameterized.named_parameters(
+        {
+            TESTCASE_NAME: f"test_pmi_mask_word",
+            EXAMPLES: DEMO_TEXTS,
+            PMI_VOCAB: PMI_DEMO_VOCAB,
+        },
+    )
```
Comment (zbambergerNLP) [inline pmi_corruption_test.py]: Make a unit test (i.e., a named parameter dictionary) for each of the tests above. It should be clear what is happening in each unit test.

File: pre_train_t5.py [hunk 15]
```diff
@@ -29,11 +29,14 @@ def main():
     accelerator.print('training_args:', training_args)
     accelerator.print('model_args:', model_args)
     accelerator.print('data_args:', data_args)
-
     accelerator.print(f'Started pre-training a {model_args.model_name_or_path} model')
+    os.environ["WANDB_API_KEY"] = '0bc4005546f2e4d5791ff50a0eeb476ee687d5c6'
```
Comment (zbambergerNLP) [inline pre_train_t5.py:33]: Remove this

**Review by zbambergerNLP — COMMENTED** [review]
Please see my comments in this review. Great start!

File: corruption_t5.py (hunk 6 above)
Comment (OfekGlick) [reply corruption_t5.py:125]: Can you expand on this point?
From analyzing the way the T5 tokenizing works, most words have their own specific token. Where every token (other then the first word) will have a "_" before the word itself. 
If a word is comprised of a number of tokens, the first token will have a "_" before it, while the rest of the tokens will not. So I believe these "if" statemesnts cover 99% of cases.

File: corruption_t5_test.py
Comment (OfekGlick) [reply corruption_t5_test.py:1]: Can you expand on this point?

File: pmi_corruption_test.py (hunk 14 above)
Comment (OfekGlick) [reply pmi_corruption_test.py]: Can you expand on this point as well? what is not clear about what is happening?

Comment (zbambergerNLP) [reply pmi_corruption_test.py]: See an example of the format I am looking for: 

```
@parameterized.named_parameters(
        {
            'testcase_name': 'short post',
            'input_argument': 'Israel is a terrorist state.',
            'is_pro_israel': False,
        },
        {
            'testcase_name': 'another short post',
            'input_argument': 'Israel is an apartheid state.',
            'is_pro_israel': False,
        },
        {
            'testcase_name': 'medium length post',
            'input_argument': 'Hamas was justified. What else can Palestinians do in the face of occupation?',
            'is_pro_israel': False,
        },
        {
            'testcase_name': 'short supportive',
            'input_argument': 'In conflicts, civilian populations often suffer the most, highlighting the need for '
                              'peaceful resolution and humanitarian aid.',
            'is_pro_israel': True,
        },
        {
            'testcase_name': 'medium supportive',
            'input_argument': 'The historical and ongoing conflict in the Middle East involves complex geopolitical '
                              'and social factors, making simplistic labels misleading.',
            'is_pro_israel': True,
        },
        {
            'testcase_name': 'long supportive',
            'input_argument': 'It is crucial to consider the perspectives of international law and human rights '
                              'organizations in understanding the dynamics of the Israel-Palestine conflict.',
            'is_pro_israel': True,
        }
    )
    def test_generate_response(
            self,
            input_argument,
            is_pro_israel,
            audience=src_persuade.Audience.MODERATE,
            counterargument_length=src_persuade.Length.FEW_SENTENCES,

    ):
```
    
    The point of creating a dictionary for each unit test is that it's very easy to then pinpoint which test went wrong, and how that test was configured.

File: corruption_t5.py (hunk 6 above)
Comment (zbambergerNLP) [reply corruption_t5.py:125]: You make a good point about sentence-piece tokens starting with an underscore. I think we can resolve this comment. Let's exercise caution when we perform tokenization (break down words into tokens) on the basis of spaces, since we'd also need to account for other forms of spaces such as tabs and linebreaks. If the logic is determined by the common underscore that precedes words in sentence-piece, then things are good by my standards.

File: corruption_t5_test.py
Comment (zbambergerNLP) [reply corruption_t5_test.py:1]: The unit tests you created in pmi_corruption_test.py should be here. Each new function you created in corruption_t5.py should be tested in this file in particular. You should identify edge cases and test against them as well (e.g., what happens if you encounter the empty string, vs a 1-line text, vs. a multi-line text with different spacing types).

**Review by zbambergerNLP — COMMENTED** [review]
Answered your questions and clarifications

File: corruption_t5.py [hunk 16]
```diff
@@ -43,7 +43,27 @@ def shift_tokens_right(
...
-def merge_segments(offset, n_grams, covered_indices, whole_words, start_inds, segments_to_merge, ngrams_vocab_set):
+def merge_segments(
+        offset: int,
+        n_grams: int,
+        covered_indices: Set[int],
+        whole_words: str,
+        start_inds: List[int],
+        segments_to_merge: List[List[int]],
+        ngrams_vocab_set: Set[str],
+):
```
Comment (zbambergerNLP) [inline corruption_t5.py:54]: Missing return type

File: corruption_t5.py [hunk 17]
```diff
@@ -58,32 +78,41 @@ def merge_segments(offset, n_grams, covered_indices, whole_words, start_inds, se
         segments_to_merge.append(seg_inds)
 
 
-def pmi_word_mask(input_tokens, pmi_vocab, max_predictions=512, mlm_probability=0.5):
+def pmi_word_mask(
+        input_tokens: List[str],
+        pmi_vocab: Set[str],
+        max_predictions=512,
+        mlm_probability=0.5,
+):
```
Comment (zbambergerNLP) [inline corruption_t5.py]: Missing return type

File: corruption_t5.py [hunk 18]
```diff
@@ -43,7 +43,27 @@ def shift_tokens_right(
...
+    Merge segments of length n_grams into a single segment if they are in the ngrams_vocab_set.
+    Args:
+        offset: The offset of the current segment in the whole sequence.
+        n_grams: The length of the segment to merge.
+        covered_indices: A set of indices that were already merged.
+        whole_words: The whole sequence.
+        start_inds: The indices of the start of each word in the whole sequence.
+        segments_to_merge: A list of segments to merge.
+        ngrams_vocab_set: The set of ngrams to use for PMI-based corruption.
+    """
```
Comment (zbambergerNLP) [inline corruption_t5.py:65]: Missing description of returned object

File: corruption_t5.py [hunk 19]
```diff
@@ -58,32 +78,41 @@ def merge_segments(offset, n_grams, covered_indices, whole_words, start_inds, se
...
     """
     Create a mask for PMI-based corruption for a sample.
+    Initially, we map which tokens are part of the same word, and then we mask the entire word.
+    Next, we mask ngrams that are in the ngrams_vocab_set, from 5 to 2 grams, while avoiding overlapping ngrams.
     Args:
         input_tokens: A tensor of tokens.
         pmi_vocab: The set of ngrams to use for PMI-based corruption.
         max_predictions: The maximum number of tokens to mask.
         mlm_probability: The probability of masking a token.
     """
```
Comment (zbambergerNLP) [inline corruption_t5.py:98]: Missing description of returned object

File: corruption_t5.py [hunk 20]
```diff
@@ -500,6 +529,7 @@ def corrupt_for_vanilla_t5(
         pmi: Whether to use PMI-based corruption.
         ngram_vocab_set: The set of ngrams to use for PMI-based corruption.
         tokenizer: The tokenizer to use for PMI-based corruption.
+        seed: The seed to use for random number generation.
```
Comment (zbambergerNLP) [inline corruption_t5.py]: No need for the seed in this part of the code. Within the test, set the seed during 'setup()', or at the beginning of the unit test. You can even pass in a seed as part of a parameterized unit tests. You should not set the seed or deal with logic surrounding it within the function itself.

File: corruption_t5.py [hunk 21]
```diff
@@ -515,7 +545,7 @@ def corrupt_for_vanilla_t5(
     # TODO: Fix the logic below given that the attention mask consists of a list of tensors of rank 2,
     #  and not a single tensor of rank 3.
     batch['attention_mask'] = torch.asarray(batch['attention_mask'])
-
+    random.seed(seed)
```
Comment (zbambergerNLP) [inline corruption_t5.py]: Drop this here, and add it to the appropriate test function

File: corruption_t5_test.py [hunk 22]
```diff
@@ -254,6 +360,94 @@ def test_corrupt_for_vanilla_t5(
...
+        ref_tokens = []
+        for input_id in input_tokens:
+            token = self.tokenizer._convert_id_to_token(input_id.item())
+            ref_tokens.append(token)
+        mask_labels_for_sample = corruption_lib.pmi_word_mask(
+            ref_tokens,
+            PMI_DEMO_VOCAB,
+            max_predictions,
+            mlm_probability, )
+        self.assertIsNotNone(mask_labels_for_sample)
```
Comment (zbambergerNLP) [inline corruption_t5_test.py:440]: Make more complex assumptions. Since you set the seed for each of the unit tests, you can always expect the same results. Make sure that you keep track of what the result should be, make sure it makes sense to you, and then compare the tensor you get from the function with the one you expect (as passed in via parameterized unit tests).

File: corruption_t5_test.py [hunk 23]
```diff
@@ -254,6 +360,94 @@ def test_corrupt_for_vanilla_t5(
...
+            add_special_tokens=False,
+            truncation=True,
+            padding=True,
+        )
+        predicted_mask_labels = corruption_lib.pmi_noise_mask(
+            tokenized_examples,
+            pmi_vocab,
+            self.tokenizer,
+        )
+        self.assertIsNotNone(predicted_mask_labels)
```
Comment (zbambergerNLP) [inline corruption_t5_test.py:474]: See my comment above about more robust unit testing.

**Review by zbambergerNLP — COMMENTED** [review]
Nice progress! The core issue at this point is to make the unit tests more robust.

File: corruption_t5.py (hunk 16 above)
Comment (OfekGlick) [reply corruption_t5.py:54]: Nothing is to be returned

File: corruption_t5.py (hunk 21 above)
Comment (OfekGlick) [reply corruption_t5.py]: Why? since I am using `random` for my seeding, unless I do it in this instance the random calls will not be deterministic.

File: corruption_t5.py (hunk 20 above)
Comment (OfekGlick) [reply corruption_t5.py]: I feel like this should depend on the way we decide to seed.
https://stackoverflow.com/questions/65685060/unique-seed-acrossing-multiple-imported-files-with-random-module-python
According to this answer, in the native `random` module, for random calls to be deterministic across different files, you need to pass the random seed. To avoid that, you can just set it at each file.

To avoid it all together, the `random` module should not be used and should be switched for something else

File: corruption_t5.py (hunk 21 above)
Comment (OfekGlick) [reply corruption_t5.py]: See my answer above

Comment (zbambergerNLP) [reply corruption_t5.py]: See CorruptionTest in `corrruption_t5_test.py`. You'll see there that if you set the seed at the start of the test, you will get deterministic behavior. In those unit tests, we compare corruption functions on specific tensor inputs, and expect to get back specific tensor outputs.

File: corruption_t5.py (hunk 20 above)
Comment (zbambergerNLP) [reply corruption_t5.py]: See CorruptionTest in corrruption_t5_test.py. You'll see there that if you set the seed at the start of the test, you will get deterministic behavior. In those unit tests, we compare corruption functions on specific tensor inputs, and expect to get back specific tensor outputs.

https://github.com/zbambergerNLP/ml_training/blob/master/corruption_t5_test.py#L39-L91

**Comment by zbambergerNLP** [comment]
I think we're nearly there. I responded to your recent comments.

File: corruption_t5_test.py (hunk 23 above)
Comment (OfekGlick) [reply corruption_t5_test.py:474]: This method doesn't have many varied uses. I don't know what other reasonable cases to test

**Review by zbambergerNLP — APPROVED** [review]
Looks great! Awesome work :D
